sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java

```diff
@@ -472,16 +475,72 @@ public <T> CosmosPagedFlux<T> queryItems(String query, Class<T> classType) {
+            ProactiveContainerInitConfig proactiveContainerInitConfig = new ProactiveContainerInitConfigBuilder(Arrays.asList(cosmosContainerIdentity))
+                    .setProactiveConnectionRegions(1)
+                    .build();
+
+            return withContext(context -> openConnectionsAndInitCachesInternal(proactiveContainerInitConfig)
                                             .flatMap(openResult -> {
                                                 logger.info("OpenConnectionsAndInitCaches: {}", openResult);
                                                 return Mono.empty();
                                             }));
         } else {
-            logger.warn(
-                    String.format(
-                        "OpenConnectionsAndInitCaches is already called once on Container %s, no operation will take place in this call",
-                        this.getId()));
+            logger.warn("OpenConnectionsAndInitCaches is already called once on Container {}, no operation will take place in this call", this.getId());
+            return Mono.empty();
+        }
+    }
+
+    /**
+     *  Best effort to initialize the container by warming up the caches and connections to a specified no. of regions from the
+     *  preferred list of regions.
+     *
+     *  Depending on how many partitions the container has, the total time needed will also change. But generally you can use the following formula
+     *  to get an estimated time:
+     *  If it took 200ms to establish a connection, and you have 100 partitions in your container
+     *  then it will take around (100 * 4 / CPUCores) * 200ms to open all connections after get the address list
+     *
+     *  <p>
+     *  <br>NOTE: This API ideally should be called only once during application initialization before any workload.
+     *  <br>In case of any transient error, caller should consume the error and continue the regular workload.
+     *  </p>
+     * <p>
+     * In order to minimize latencies associated with warming up caches and opening connections
+     * the no. of proactive connection regions cannot be more than {@link ProactiveContainerInitConfigBuilder#MAX_NO_OF_PROACTIVE_CONNECTION_REGIONS}.
+     * </p>
+     *
+     * @param numProactiveConnectionRegions the no of regions to proactively connect to
+     * @return Mono of Void.
+     */
+    public Mono<Void> openConnectionsAndInitCaches(int numProactiveConnectionRegions) {
+
+        List<String> preferredRegions = clientAccessor.getPreferredRegions(this.database.getClient());
+        boolean endpointDiscoveryEnabled = clientAccessor.isEndpointDiscoveryEnabled(this.database.getClient());
+
+        checkArgument(numProactiveConnectionRegions >= 1, "no. of proactive connection regions should be greater than 1");
+
+        if (numProactiveConnectionRegions > 1) {
+            checkArgument(endpointDiscoveryEnabled, "endpoint discovery should be enabled when no. " +
```

Comment:
after a second thought, I am wondering whether `requires preferred regions to be configured` is too strict -> if the customer does not configure preferred regions, why not just open based on the default region list order.

And would it be considered as breaking change for openConnectionsAndInitCaches() -> previously it does not require preferredRegions to be configured, now it will throw an exception

Reply:
+1 on avoiding breaking change. Whether or not to require preferredRegions? I am for enforcing it - effctively we don't have anything else to make good decisions. So for 1 region don't enforce it - but if you want to enable > 1 leave the enforcement. But let's discuss in Java review meeting.

Reply:
The >= 1 check was unintentional. Resolving this.